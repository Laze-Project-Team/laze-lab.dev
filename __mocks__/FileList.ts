class FileList<T> {
  _files: T[];
  length: number;
  constructor(files: T[]) {
    this._files = files;
    this.length = this._files.length;
  }
  item(index: number) {
    return this._files[index];
  }
  get [Symbol.toStringTag]() {
    return 'FileList';
  }
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
window.FileList = FileList;

export {};
