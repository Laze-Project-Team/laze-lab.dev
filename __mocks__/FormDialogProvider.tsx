import type { FC, ReactNode } from 'react';
import { useForm } from 'react-hook-form';

import { FormDialogContextProvider } from '@/components/ui/FormDialog/FormDialogContext';

export const FormDialogProvider: FC<{ children?: ReactNode }> = ({
  children,
}) => {
  const { control } = useForm();
  return (
    <FormDialogContextProvider
      value={{
        id: 'mock-form-dialog',
        control,
        onSubmit: async (data) => console.log(data),
      }}
    >
      {children}
    </FormDialogContextProvider>
  );
};
