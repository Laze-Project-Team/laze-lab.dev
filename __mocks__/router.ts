import type { useRouter } from 'next/router';
import { beforeEach, vi } from 'vitest';

export const mockRouter = (): void => {
  vi.mock('next/router', () => ({
    useRouter: () => {
      const router: ReturnType<typeof useRouter> = {
        asPath: '/',
        back: vi.fn(),
        basePath: '/',
        beforePopState: vi.fn(),
        events: {
          on: vi.fn(),
          off: vi.fn(),
          emit: vi.fn(),
        },
        isFallback: false,
        isLocaleDomain: false,
        isPreview: false,
        isReady: true,
        locale: 'en',
        locales: ['en'],
        prefetch: vi.fn(),
        push: vi.fn(),
        query: {},
        reload: vi.fn(),
        replace: vi.fn(),
        route: '/',
        pathname: '/',
      };

      return router;
    },
  }));
};

export const mockEachRouter = (): void => {
  beforeEach(() => {
    mockRouter();
  });
};
