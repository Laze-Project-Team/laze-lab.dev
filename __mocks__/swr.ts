import type { SWRResponse } from 'swr';

export const mockSWRResponse = <T>(data: T): SWRResponse<T> => ({
  isValidating: false,
  mutate: async () => void 0,
  data,
});
