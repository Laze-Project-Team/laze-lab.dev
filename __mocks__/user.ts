import type { User } from 'firebase/auth';
import type { SWRResponse } from 'swr';

import type { userData } from '@/typings/database';

export const userMock: User = {
  uid: 'unique-id',
  delete: async () => void 0,
  displayName: 'Laze Project Team',
  email: 'laze@example.com',
  emailVerified: true,
  getIdToken: async () => 'token',
  getIdTokenResult: async () => ({
    authTime: new Date().toISOString(),
    expirationTime: new Date().toISOString(),
    issuedAtTime: new Date().toISOString(),
    signInProvider: null,
    signInSecondFactor: null,
    token: 'token',
    claims: {},
  }),
  isAnonymous: false,
  metadata: {},
  phoneNumber: null,
  photoURL: null,
  providerData: [],
  providerId: 'email',
  refreshToken: 'token',
  reload: async () => void 0,
  tenantId: null,
  toJSON: () => ({}),
};

export const userDataMock: SWRResponse<userData | null | undefined> = {
  isValidating: false,
  mutate: async (_a: unknown, _b: unknown) => void 0,
  data: {
    name: 'Laze Project Team',
    avatarURL: 'https://laze.ddns.net/img/logo/logo.png',
    locale: 'ja',
    projects: [],
    createdAt: new Date(),
  },
};

export const getUserDataMock = <T extends Partial<userData>>(
  data: T,
): SWRResponse<userData | null | undefined> => ({
  isValidating: false,
  mutate: async (_a: unknown, _b: unknown) => void 0,
  data: {
    name: 'Laze Project Team',
    avatarURL: 'https://laze.ddns.net/img/logo/logo.png',
    locale: 'ja',
    projects: [],
    createdAt: new Date(),
    ...data,
  },
});

export const userDataLoadingMock: SWRResponse<userData | null | undefined> = {
  isValidating: false,
  mutate: async (_a: unknown, _b: unknown) => void 0,
  data: null,
};
