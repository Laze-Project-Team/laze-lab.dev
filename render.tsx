import type {
  Queries,
  queries,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { render as defaultRender } from '@testing-library/react';
import type { FC, ReactNode } from 'react';

import { ColorModeProvider } from '@/components/contexts/ColorModeContext';

export type renderType<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container,
> = (
  ui: React.ReactElement,
  options?: RenderOptions<Q, Container, BaseElement>,
) => RenderResult<Q, Container, BaseElement>;

type wrapperProps = {
  children?: ReactNode;
};

type globalWrapperHOCType = (
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>,
) => FC<wrapperProps>;

const globalWrapperHOC: globalWrapperHOCType = (Wrapper) => {
  const globalWrapper: FC<wrapperProps> = ({ children }) => (
    <ColorModeProvider>
      {Wrapper ? (
        <Wrapper>
          <>{children}</>
        </Wrapper>
      ) : (
        children
      )}
    </ColorModeProvider>
  );

  return globalWrapper;
};

export const render: renderType = (ui, options) =>
  defaultRender(ui, {
    ...options,
    wrapper: globalWrapperHOC(options?.wrapper),
  });
