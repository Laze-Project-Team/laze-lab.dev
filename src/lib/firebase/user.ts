import type { User } from 'firebase/auth';
import { onAuthStateChanged } from 'firebase/auth';
import type { UpdateData } from 'firebase/firestore';
import { deleteDoc, doc, getDoc, setDoc, updateDoc } from 'firebase/firestore';
import { getDownloadURL, ref, uploadBytes } from 'firebase/storage';
import { useEffect } from 'react';
import type { SWRResponse } from 'swr';
import { mutate } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { timestampToDate } from '@/lib/firebase/TimestampToDate';
import { isLocale } from '@/lib/utils/isLocale';
import type { baseUserData, userData } from '@/typings/database';

import { useFirebase } from '.';

export type userManager = {
  getUser: () => SWRResponse<User | null | undefined>;
  syncUser: () => void;
  createUser: (user: User) => Promise<userData>;
  fetchUserData: (user: User) => Promise<userData>;
  updateUserData: (user: User, data: UpdateData<baseUserData>) => Promise<void>;
  updateUserAvatar: (user: User, avatarFile: File) => Promise<void>;
  deleteUser: (user: User) => Promise<void>;
};

export const useUserManager = (): userManager => {
  const { auth, db, storage } = useFirebase();
  const currentUser = useSWRImmutable<User | null | undefined>(
    'currentUser',
    () => auth.currentUser,
  );

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      mutate('currentUser', user);
    });

    return unsubscribe;
  }, [auth]);

  const getUser: userManager['getUser'] = () => {
    return currentUser;
  };

  const syncUser: userManager['syncUser'] = () => {
    mutate('currentUser', undefined);
    mutate('currentUser', auth.currentUser);
  };

  const createUser: userManager['createUser'] = async (user) => {
    const userData: userData = {
      avatarURL: user.photoURL,
      locale: isLocale(auth.languageCode) ? auth.languageCode : 'en',
      name: user.displayName ?? 'user',
      projects: [],
      createdAt: new Date(),
    };

    await setDoc(doc(db, 'users', user.uid), userData);

    return userData;
  };

  const fetchUserData: userManager['fetchUserData'] = async (user) => {
    const snapshot = await getDoc(doc(db, 'users', user.uid));

    if (!snapshot.exists()) {
      return await createUser(user);
    }

    const userData = timestampToDate(snapshot.data()) as userData;

    return userData;
  };

  const updateUserData: userManager['updateUserData'] = async (user, data) => {
    await updateDoc(doc(db, 'users', user.uid), data);
  };

  const updateUserAvatar: userManager['updateUserAvatar'] = async (
    user,
    avatarFile,
  ) => {
    const avatarFileRef = ref(storage, `users/${user.uid}`);

    const { ref: uploadedRef } = await uploadBytes(avatarFileRef, avatarFile);

    const avatarURL = await getDownloadURL(uploadedRef);

    await updateUserData(user, { avatarURL });
  };

  const deleteUser: userManager['deleteUser'] = async (user) => {
    await deleteDoc(doc(db, 'users', user.uid));
  };

  return {
    getUser,
    syncUser,
    createUser,
    fetchUserData,
    updateUserData,
    updateUserAvatar,
    deleteUser,
  };
};
