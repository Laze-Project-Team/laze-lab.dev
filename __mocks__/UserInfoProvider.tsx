import type { FC, ReactNode } from 'react';
import { vi } from 'vitest';

import { mockSWRResponse } from '/__mocks__/swr';
import { userDataMock, userMock } from '/__mocks__/user';
import { userInfoContext } from '@/components/contexts/UserInfoContext';

export const syncUserMock = vi.fn();
export const syncUserDataMock = vi.fn();

export const UserInfoProvider: FC<{ children?: ReactNode }> = ({
  children,
}) => (
  <userInfoContext.Provider
    value={{
      user: mockSWRResponse(userMock),
      userData: mockSWRResponse(userDataMock),
      syncUser: syncUserMock,
      syncUserData: syncUserDataMock,
    }}
  >
    {children}
  </userInfoContext.Provider>
);
