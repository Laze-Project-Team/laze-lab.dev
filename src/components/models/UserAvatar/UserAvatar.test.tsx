import '@testing-library/jest-dom';

import { expect, it } from 'vitest';

import { userDataMock, userMock } from '/__mocks__/user';
import { render } from '@/lib/test/render';

import { UserAvatar } from '.';

it('should be rendered correctly', () => {
  const { container } = render(
    <UserAvatar user={userMock} userData={userDataMock} />,
  );
  expect(container.firstChild).toMatchSnapshot();
});
