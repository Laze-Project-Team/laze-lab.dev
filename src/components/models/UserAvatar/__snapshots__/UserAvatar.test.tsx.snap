// Vitest Snapshot v1

exports[`should be rendered correctly 1`] = `
<div
  class="css-4r7jvw-PresentialUserAvatar"
>
  <div
    class="MuiAvatar-root MuiAvatar-circular css-sxq756-MuiAvatar-root"
  >
    <img
      alt="Laze Project Team"
      class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
      src="https://laze.ddns.net/img/logo/logo.png"
    />
  </div>
  <label
    aria-label="profile.avatar.edit.title"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1ssy60s-MuiButtonBase-root-MuiIconButton-root-UploadAvatarButton-PresentialUserAvatar"
    data-mui-internal-clone-element="true"
    role="button"
    tabindex="0"
  >
    <input
      accept="image/*"
      hidden=""
      type="file"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="FileUploadIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </label>
</div>
`;
